Some refactor to make the code cleaner

ecastro98 opened the pull request:
It would also be able to be used with composer, the package manager for php.

ecastro98 committed:
Reorganize code

ecastro98 committed:
Add composer and apply PSR

ecastro98 committed:
Change filename to its name class

ecastro98 committed:
Split classes and fix getInput function

ecastro98 committed:
Execute php-cs-fixer for cleaner code

ecastro98 committed:
Make code shorter

ecastro98 committed:
Add sdk new version

ecastro98 committed:
Add Travis-CI

qaisjp commented:
I've deleted commit 0205974 & fixed some other bugs to make the commit history cleaner. You will need to `git fetch` and then `git reset origin/feature/refactor --hard` so that you get the updated repo.

ecastro98 committed:
Add credential class

ecastro98 committed:
Add server class

ecastro98 committed:
Add resources class

ecastro98 committed:
Add json and curl to composer
Add server implementation to ajax

ecastro98 committed:
Use guzzle library for http calls

ecastro98 committed:
Uppercase Mta class

shyim commented on src/Model/Resources.php:
```diff
@@ -0,0 +1,41 @@
+
+    public function add(Resource $resource): bool
+    {
+        return (bool) $resource[] = $resource;
```
Maybe void?

shyim commented on composer.json:
```diff
@@ -0,0 +1,47 @@
+        "php": "^7.0",
+        "ext-curl": "^7.2",
+        "ext-json": "^1.6",
+        "guzzlehttp/guzzle": "^6.3"
```
Maybe use a client abstraction like http://httplug.io/ ?

shyim commented on src/Model/Resource.php:
```diff
@@ -0,0 +1,49 @@
+
+    public function call($function, ...$arguments)
+    {
+        if (!$this->server) {
```
Maybe make $server in constructor instead required?

ecastro98 replied:
This is because getInput, static method of Mta class, use the constructor. That's why I made it optional

shyim commented on src/Model/Resource.php:
```diff
@@ -0,0 +1,49 @@
+    private $name;
+    private $server;
+
+    public function __construct($name, $server = null)
```
Typehintings are missing :)

shyim commented on composer.json:
```diff
@@ -0,0 +1,47 @@
+    },
+    "autoload-dev": {
+        "psr-4": {
+            "MultiTheftAuto\\Sdk\\": "tests/"
```
Autoload Namespace is registered two times on dev mode :)

ecastro98 committed:
Fix php-cs-fixer and phpstan

ecastro98 committed:
Resolve @shyim reviews

shyim commented on composer.json:
```diff
@@ -0,0 +1,43 @@
+    "minimum-stability": "dev",
+    "prefer-stable": true,
+    "require": {
+        "php": "^7.0",
```
This should be increased to 7.1.
You have used https://github.com/multitheftauto/mtasa-php-sdk/pull/2/files#diff-74763f9c84d84cc78a22ed38d80d45f9R25 nullable return types, which are introduced with PHP 7.1 :)

ecastro98 replied:
Yeah, I was trying to make it compatible with PHP 7, didn't know about that. I will change it, thanks

Sasu98 replied:
Or is it better to replace with false boolean? I mean, there is an important number of people using PHP 7. What do you think?

shyim replied:
I would increase just the php version to 7.1 and done.
PHP 7.0 security support already left out 😅

ecastro98 committed:
Change PHP Version require to ^7.1

ecastro98 commented:
For some reason, WIP check dissapeared. Do not merge while 'WIP' is in title.

ecastro98 committed:
Fix typo

ecastro98 committed:
Change SDK Version to 1.0.0
This version is not compatible with version 0.5 so that MAJOR version change is required. See more at https://semver.org/

ecastro98 committed:
Add translator class

ecastro98 committed:
Add RequestFactory and HttpStatusVerification classes

ecastro98 committed:
Fix strict return type in Translator

ecastro98 committed:
Fix phpstan

ecastro98 committed:
Add break lines between parameters

ecastro98 committed:
Make translator do json stuff
Fix do_post_request not returning string

ecastro98 committed:
Add missing return types in Mta class

ecastro98 committed:
Add Input class for catch data from mta server

ecastro98 committed:
Make util classes not instantiable

ecastro98 committed:
Add tests cases

ecastro98 committed:
Fix phpstan

ecastro98 committed:
Remove project version from files
It will better to use git tag

ecastro98 committed:
Change variable's name to its respective class type

ecastro98 committed:
Fix expected exception message in resource test

ecastro98 committed:
Add one more test case

ecastro98 committed:
Fix test

ecastro98 committed:
Add some information to README

ecastro98 committed:
Fix path using double slash

ecastro98 committed:
Fix redefining variable as request is immutable

ecastro98 commented:
I think that is all for now. Thanks @shyim for your feedback. Can you please review this @qaisjp and @Deltanic ?

ecastro98 committed:
Add method magic to direct call in Resource

ecastro98 committed:
Fix testItReturnsBodyFromStream and ignore getInput test

qaisjp commented:
So much stuff! Awesome work. Heads up: it **will** take us a long time to merge this pull request.

ecastro98 committed:
Add magic getter to have another way for getResource

ecastro98 committed:
Show how to pass arguments

ecastro98 committed:
fix: arguments using magic method were passed as array

ecastro98 committed:
refactor: throw exception if resource return not found error

qaisjp commented on examples/ajax/ajax_interface.php:
```diff
@@ -0,0 +1,40 @@
+
+	$server = new Server($host, $port);
+	$credential = new CredentialTest($http_username, $http_password);
+    $mta = new Mta($server, $credential);
```
heads up

ecastro98 committed:
refactor: some design changes

ecastro98 committed:
build: fix new cmd name

ecastro98 committed:
style: add no-unused-imports rule

ecastro98 committed:
refactor: implements more abstraction

ecastro98 committed:
refactor: cast to string response body

ecastro98 committed:
fix: reassign request with body

ecastro98 committed:
fix: update example

ecastro98 committed:
fix: remove composer.lock from gitignore

ecastro98 committed:
ci: remove travis

ecastro98 committed:
ci: add azure pipeline

[skip ci]

ecastro98 committed:
build: set minimum php version to 7.3

ecastro98 committed:
style: change travis badge to azure one

ecastro98 commented:
More and more changes >:D @qaisjp @Deltanic

qaisjp committed:
doc: tweak readme grammar

qaisjp reviewed (approved):
I gave it a read, this looks great!

Whilst I cannot judge whether or not the code is idiomatic (I am not a PHP developer), I couldn't find any obvious issues 👍 

My only comment would be that I am not sure if supporting the following syntax is a good idea:

```php
$response = $mta->someResource->callableFunction($arg1, $arg2, $arg3, ...);
```

My concern is that, if we add more ways to interact with a server, the possibility of clashing with existing resources might prevent us from adding more methods.

What do you think? It is totally your call whether or not to keep the functionality.

ecastro98 committed:
refactor: remove magic getter

ecastro98 commented:
We have MtaService class where it happens the real interaction with the server. The magic method (in Resource class) will execute the method `call` from the service so if we add more ways to interact with a server, it should happen here with no problems.

Plus, it also adds some DX.

Although, I will remove the magic getter from Mta class and keep `$mta->getResource('someResource')` to get a resource because it restrict us to create public properties that could be important in the future.

So the final ways to call a resource function will be:
```php
$response = $mta->getResource('someResource')->call('callableFunction', $arg1, $arg2, $arg3, ...);
// or
$response = $mta->getResource('someResource')->callableFunction($arg1, $arg2, $arg3, ...);
```

qaisjp commented:
> Although, I will remove the magic getter from Mta class and keep `$mta->getResource('someResource')` to get a resource because it restrict us to create public properties that could be important in the future.

This is actually what I was referring to. I didn't even notice the `$res->callableFunction(...)` shorthand!

I notice `Resource.php` has this:

```php
public function getName()
{
    return $this->name;
}
```

If there's a callable function `getName`, wouldn't you be required to use `$resource->call('getName', ...)` instead?

I agree it's nice to have the shorthand, but imo there's the same issue here. Maybe a nice way around that would be to support `$res->call->callableFunction(...)` or `$res->rpc->callableFunction(...)` instead?

ecastro98 commented:
Agree. I will add it tomorrow.

qaisjp commented:
Oh, and one more thing, the license.

It looks like all of this is original work, so the license should be completely up to you.

I'd recommend either:
- [GNU AGPL v3](https://github.com/multitheftauto/mtasa-php-sdk/community/license/new?branch=master&template=agpl-3.0) ([choosealicense.com](https://choosealicense.com/licenses/agpl-3.0/)), or
- [MIT](https://github.com/multitheftauto/mtasa-php-sdk/community/license/new?branch=master&template=mit) ([choosealicense.com](https://choosealicense.com/licenses/mit/))

ecastro98 committed:
docs: add license

ecastro98 committed:
docs: update php version requirement in readme

ecastro98 committed:
refactor: add new way to make function calls

ecastro98 commented:
@qaisjp What about now?

qaisjp reviewed (approved):
Go for merge!
